### PR TITLE
[web] Render RTL text correctly

### PR DESCRIPTION
**Pull request by mdebbar**
Implement handling of bidi text in the html renderer.

*More details and explanations coming soon*

Goldens: https://github.com/flutter/goldens/pull/205
Golden updates: https://github.com/flutter/goldens/pull/210

Fixes https://github.com/flutter/flutter/issues/69396
Fixes https://github.com/flutter/flutter/issues/32239
Fixes https://github.com/flutter/flutter/issues/78550
Fixes https://github.com/flutter/flutter/issues/82136

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 1]
```diff
@@ -516,41 +590,111 @@ class SpanBox extends RangeBox {
...
+      //              +---------------------+
+      //              | L o r e m i p s u m |
+      //              +---------------------+
+      //    this.left ^                     ^ this.right
+      left = this.left + before;
+      right = this.right - after;
+    } else {
+      // Example: let's say the text is "txet_werbeH" ("Hebrew_text" flowing from
+      // right to left). Say we want to get the box for "brew". The `before` is
+      // the width of "He", and `right` is the width of "_text".
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:629]:
```suggestion
      // the width of "He", and `after` is the width of "_text".
```

File: lib/web_ui/test/text/layout_service_plain_test.dart [hunk 2]
```diff
@@ -41,147 +41,147 @@ void main() {
...
-    expect(paragraph.maxIntrinsicWidth, 110);
-    expect(paragraph.minIntrinsicWidth, 30);
-    expect(paragraph.width, 70);
-    expect(paragraph.height, 20);
-    expectLines(paragraph, [
-      l('foo bar ', 0, 8, hardBreak: false, width: 70.0, left: 0.0, height: 10.0, baseline: 8.0),
-      l('baz', 8, 11, hardBreak: true, width: 30.0, left: 0.0, height: 10.0, baseline: 18.0),
-    ]);
-  });
+  // test('no text', () {
```
Comment (yjbanov) [inline lib/web_ui/test/text/layout_service_plain_test.dart:44]: What's the plan for these tests?

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 3]
```diff
@@ -367,35 +436,21 @@ abstract class RangeBox {
   ///
   /// The [x] offset is expected to be relative to the left edge of the line,
   /// just like the coordinates of this box.
-  ui.TextPosition getPositionForX(double x);
+  ui.TextPosition getPositionForX(double x, EngineLineMetrics line);
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:439]: @yjbanov which part did you expect to be updated? I think the dartdocs are still valid even after removing the `line` parameter.

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 4]
```diff
@@ -1081,9 +1247,108 @@ class LineBuilder {
     );
   }
 
+  /// Position the boxes taking into account bi-directional text, and the
+  /// paragraph's direction.
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:1251]: That's what it does. It positions the boxes :)

File: lib/web_ui/lib/src/engine/text/text_direction.dart [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+  final LineBreakResult lineBreak;
+
+  final ui.TextDirection? textDirection;
+
+  final bool isSpaceOnly;
+
+  LineBreakType get type => lineBreak.type;
+
+  /// Creates a copy of this [DirectionalPosition] with a different [index].
```
[the remark this answers is not in the document]
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/text_direction.dart:59]: By default, we don't want to break anywhere, except where `nextLineBreak()` says it's ok to break.

File: lib/web_ui/test/text/layout_service_plain_test.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/test/text/layout_service_plain_test.dart:44]: Whoops! They should be fine. I just forgot to uncomment them.

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/text/layout_service.dart]: Ah, sorry, I forgot to remove this comment after we decided that `line` is no longer needed.

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (yjbanov) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:1255]: Hmm, I remember adding this comment, but I don't recall it being related to this method. I must have misclicked, adding the comment elsewhere. I saw a method that jumped straight into describing its parameters without saying what it does. I can't find it any more 🤷‍♂️ Oh well...

**Review by yjbanov — APPROVED** [review]
lgtm

**Pull request merged**